File: tools/modal-sweeper/modal_sweeper.py
#!/usr/bin/env python3
"""RETIRED 2026-08-04 -- MANUAL BREAK-GLASS ONLY. DO NOT run this on a timer.

  SUPERSEDED by the in-engine sweeper (src/session/modal-sweeper.ts, session-hygiene, live inside
  theoffice.service). That is the authoritative path and a strict superset of this tool: usage-limit
  dismissal (highlight-verified Stop-and-wait) PLUS permission/dangerous-op alarm, and it pages the
  owner directly. This standalone script is kept ONLY as an emergency fallback for the case where the
  in-engine sweeper itself has to be disabled. Its frame matchers MAY DRIFT from Claude Code's current
  rendering -- RE-VERIFY against a live pane before ever running it, and NEVER leave it on a timer
  (running it alongside the engine sweeper double-actuates the same modal on overlapping cadences).
  Out of git by design (see .gitignore); usage is documented in the clear-agent-modal-safely skill.

Interim usage-limit modal sweeper (Darryl, 2026-08-04).

STOPGAP for the delivery outage of 2026-08-04: when a Claude agent hits its plan limit mid-turn, Claude
Code replaces the footer with a blocking "Stop and wait / Upgrade" menu. That reads as pane-state
'unknown' to the engine, so the deliverer parks EVERY queued message behind it (the owner's included),
and it never self-clears — an unattended agent goes deaf until a human presses Enter (marveen, dwight
and toby were all frozen this way on 2026-08-04). This decoupled timer dismisses the modal by confirming
option 1, "Stop and wait for limit to reset" (Claude Code then auto-resumes at reset). It does NOT touch
the engine and is reversible: `systemctl --user disable --now modal-sweeper.timer`.

The proper in-engine fix (detector in pane-state.ts + sweep in claude-runtime) is a separate patch under
Toby's gate; this only exists so nobody freezes overnight while that lands.

GOVERNING PRINCIPLE (Michael, bus 9362): AUTOMATION MAY CONFIRM, NEVER NAVIGATE.
  Moving a cursor (a digit, an arrow) asserts "I know which option is right" — a CHOICE that requires
  reading the actual command and judging its consequences. A matcher cannot read a command; it
  pattern-matches a frame, so it is not entitled to that claim. Pressing Enter on an ALREADY-safe frame
  asserts only "the safe option is already selected" — the weaker claim a matcher IS entitled to make.
  So: if the highlight is not already where it must be, the tool has hit a state it does not understand,
  and the correct response is to alarm and leave it (detectsUnsafeUsageLimitModal), never to navigate to
  what it guesses is safe. A sweeper that navigates has promoted itself from matcher to decision-maker,
  and the first time the rendering shifts it navigates confidently to the wrong place. (Clearing a real
  permission gate by hand — Down, re-capture, confirm ❯ on "No", Enter — is a HUMAN procedure with the
  authority to choose; it must NOT be copied into this tool.)

FOUR HARD CONDITIONS (Michael, bus 9245), each enforced structurally, not by intent:

  1. STRUCTURALLY INCAPABLE OF SELECTING UPGRADE. Per the governing principle, we CONFIRM, never
     navigate. The only key ever sent is Enter, and only when the cursor glyph is provably on the
     Stop-and-wait line: `❯ 1. Stop and wait for limit to reset`. If the cursor is anywhere else (e.g. on
     option 2), the expected string is not found and we REFUSE — no keystroke at all. The real hazard was
     never "a digit" specifically; it was ACTING WITHOUT VERIFYING WHICH OPTION IS SELECTED — a bare Enter
     does that just as a stray digit does. Verified-highlight Enter closes both.
  2. NEVER DISMISS MID-TURN. A pane showing any busy signal (esc-to-interrupt, the `(Ns · ↓` token
     counter) is left completely alone — the modal only appears on a HALTED turn, so a busy pane is by
     definition not parked on it.
  3. LOG EVERY DISMISSAL AND EVERY REFUSAL — agent, ISO timestamp, what matched — to modal-sweeper.log,
     so this can never fire (or silently stop firing) unseen.
  4. Runs as a timer Toby can kill on sight.

Usage:
  modal_sweeper.py            # one sweep of all agent-* panes (what the timer runs)
  modal_sweeper.py --dry-run  # detect + log intent, send NOTHING
  modal_sweeper.py --self-test  # prove the matcher on fixtures; exits non-zero if any case is wrong
"""
import re
import os
import sys
import json
import time
import hashlib
import subprocess
import urllib.request
from datetime import datetime, timezone

TMUX = "/usr/bin/tmux"
SOCKET = "theoffice"
TENANT = os.environ.get("OFFICE_TENANT_ROOT", "/opt/claude/theoffice/tenant")
LOG = f"{TENANT}/store/modal-sweeper.log"
STATE = f"{TENANT}/store/.modal-sweeper-alarm-state.json"
API = "http://127.0.0.1:3430/api/messages"
# Re-alarm about the SAME frozen permission prompt at most this often (a human is needed, but don't spam).
PERMISSION_REALARM_SEC = 1800

# The modal signature. ALL THREE must hold for us to act:
#   (a) the cursor is on option 1 = Stop-and-wait  -> confirming it can only ever pick Stop-and-wait
#   (b) option 2 = Upgrade is present              -> confirms this really is the usage-limit menu
#   (c) the modal's own footer "Enter to confirm"  -> confirms the menu is live, not quoted scrollback
CURSOR_ON_STOP_RX = re.compile(r"❯\s*1\.\s*Stop and wait for limit to reset")
UPGRADE_RX = re.compile(r"2\.\s*Upgrade your plan")
CONFIRM_FOOTER_RX = re.compile(r"Enter to confirm")
# Busy = a turn is mid-flight; never touch it (condition 2).
BUSY_RX = re.compile(r"esc to interrupt|\(\s*(?:\d+m\s*)?\d+s\s*·\s*↓\s*\d")

# PERMISSION / APPROVAL prompt. This is the class the sweeper must be STRUCTURALLY INCAPABLE of clearing
# (Michael, bus 9248): dismissing one APPROVES an action — a delete, a push, a payment, a deploy —
# blind, on the owner's box, with his credentials. A frozen agent is an inconvenience; an unattended
# approval is unbounded. So for these we DETECT + ALARM + LEAVE UNTOUCHED. Freezing is the safe failure.
# The strings below are Claude Code's own approval-menu wording; they never appear in the usage-limit
# menu (which the classifier checks FIRST and which has "Upgrade your plan", not "Yes / No"). Agents run
# --dangerously-skip-permissions so these should be rare, which is exactly why one appearing must be
# surfaced loudly rather than left to sit silently.
PERMISSION_RX = re.compile(
    r"No, and tell Claude what to do differently"
    r"|Yes, and (?:don't|do not) ask again"
    r"|Do you want to (?:proceed|make this edit|create|run|allow)"
)


def log(msg: str) -> None:
    ts = datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds")
    line = f"{ts} {msg}"
    try:
        with open(LOG, "a") as f:
            f.write(line + "\n")
    except OSError:
        pass
    print(line)


def classify(pane: str) -> str:
    """Pure decision on a captured pane. One of:
      'act'        — the usage-limit modal is up with the cursor on Stop-and-wait; safe to send Enter.
      'refuse'     — the usage-limit modal is present BUT the cursor is not provably on Stop-and-wait.
      'permission' — a permission/approval prompt: DETECT + ALARM, NEVER a keystroke (Michael 9248).
      'busy'       — a turn is mid-flight; leave it alone.
      'none'       — nothing this tool cares about.

    Order is a safety property: usage-limit is matched by its OWN exact signature ("Upgrade your plan"
    + confirm footer), which a permission prompt never carries, so the two can never be confused. The
    ONLY verdict that yields a keystroke is 'act', and 'act' requires the cursor provably on
    Stop-and-wait — so neither Upgrade nor any permission option is ever reachable.
    """
    if BUSY_RX.search(pane):
        return "busy"
    if UPGRADE_RX.search(pane) and CONFIRM_FOOTER_RX.search(pane):
        # The usage-limit menu. Act only if the highlight is provably on Stop-and-wait.
        return "act" if CURSOR_ON_STOP_RX.search(pane) else "refuse"
    if PERMISSION_RX.search(pane):
        return "permission"
    return "none"


def _load_state() -> dict:
    try:
        with open(STATE) as f:
            return json.load(f)
    except (OSError, ValueError):
        return {}


def _save_state(state: dict) -> None:
    try:
        with open(STATE, "w") as f:
            json.dump(state, f)
    except OSError:
        pass


# Re-log a STILL-frozen permission gate at most this often. First detection always logs; per-tick
# repeats are suppressed so a long freeze doesn't bury the log (Michael 9333: a 2h40m freeze logged
# ~480 near-identical DETECTED lines). Log the TRANSITION, not the state.
DETECTED_HEARTBEAT_SEC = 1800


def note_permission_prompt(session: str, pane: str) -> None:
    """A permission prompt is up. LEAVE IT ALONE and log the TRANSITION — the first detection, then a
    periodic still-frozen heartbeat — NOT every 20s tick. Alarming is owned by the sibling detect-and-
    alarm tool (store/frozen-pane-alarm.sh, Michael 9265); routing a second alarm here would double-page
    Szoszo. The record still shows the sweeper saw it and correctly refused, just without the noise."""
    agent = session.removeprefix("agent-")
    # Key on the pane content so a genuinely NEW/different gate re-logs as a fresh transition.
    key = "det:" + hashlib.sha1((session + "|" + pane[-400:]).encode()).hexdigest()[:16]
    state = _load_state()
    now = time.time()
    last = state.get(key)
    if last is None:
        log(f"PERMISSION-PROMPT-DETECTED agent={agent} action=NONE(left untouched; alarm owned by frozen-pane-alarm.sh)")
    elif now - last >= DETECTED_HEARTBEAT_SEC:
        log(f"PERMISSION-PROMPT-STILL-FROZEN agent={agent} (~{int((now - last) / 60)}min since last note, still untouched)")
    else:
        return  # same gate, logged recently — suppress the per-tick repeat
    state[key] = now
    # prune old keys so the state file can't grow without bound
    state = {k: v for k, v in state.items() if now - v < 4 * DETECTED_HEARTBEAT_SEC}
    _save_state(state)


def list_agent_sessions() -> list[str]:
    r = subprocess.run([TMUX, "-L", SOCKET, "list-sessions", "-F", "#{session_name}"],
                       capture_output=True, text=True, timeout=10)
    if r.returncode != 0:
        return []
    return [s.strip() for s in r.stdout.splitlines() if s.strip().startswith("agent-")]


def capture(session: str) -> str | None:
    r = subprocess.run([TMUX, "-L", SOCKET, "capture-pane", "-t", session, "-p"],
                       capture_output=True, text=True, timeout=10)
    return r.stdout if r.returncode == 0 else None


def sweep(dry_run: bool) -> int:
    acted = 0
    for session in list_agent_sessions():
        pane = capture(session)
        if pane is None:
            continue
        verdict = classify(pane)
        if verdict == "act":
            if dry_run:
                log(f"DRY-RUN would-dismiss agent={session} match='❯ 1. Stop and wait for limit to reset' (Enter, Stop-and-wait)")
                acted += 1
                continue
            # Ground-truth capture (Toby 9290): record the exact matched modal lines the FIRST time we
            # act on a real one, so we have real-world proof classify() fires on the live modal — not
            # just on the fixture — and can confirm `❯` is genuinely the selection marker.
            matched = "\n".join(l for l in pane.splitlines() if ("Stop and wait for limit to reset" in l or "Upgrade your plan" in l))
            log(f"DISMISS-MATCHED-PANE agent={session} lines=<<{matched}>>")
            ok = subprocess.run([TMUX, "-L", SOCKET, "send-keys", "-t", session, "Enter"],
                                capture_output=True, text=True, timeout=10).returncode == 0
            log(f"DISMISS agent={session} modal=usage-limit action=Enter(select-Stop-and-wait) sent={ok}")
            acted += 1
        elif verdict == "refuse":
            # Loud on purpose: the modal is up but the highlight is NOT on Stop-and-wait, so acting could
            # confirm the wrong option. We refuse and record it — a refusal that needs a human eye.
            log(f"REFUSE agent={session} modal=usage-limit reason=cursor-not-on-Stop-and-wait (no key sent)")
        elif verdict == "permission":
            # NEVER auto-dismiss a permission gate: clearing it approves the action. Log + leave it;
            # the alarm is owned by frozen-pane-alarm.sh (no double-page).
            if not dry_run:
                note_permission_prompt(session, pane)
            else:
                log(f"DRY-RUN permission-prompt agent={session} action=NONE(left untouched; frozen-pane-alarm owns the alarm)")
    return acted


# -------- self-test: proves the four conditions on fixtures; CI/pre-install gate --------
def self_test() -> int:
    MODAL_CURSOR_ON_1 = "\n".join([
        "  You've hit your session limit · resets 6:20am",
        "   What do you want to do?",
        "   ❯ 1. Stop and wait for limit to reset",
        "     2. Upgrade your plan",
        "   Enter to confirm · Esc to cancel",
    ])
    MODAL_CURSOR_ON_2 = "\n".join([  # the dangerous case: highlight on Upgrade
        "   What do you want to do?",
        "     1. Stop and wait for limit to reset",
        "   ❯ 2. Upgrade your plan",
        "   Enter to confirm · Esc to cancel",
    ])
    MODAL_BUSY = MODAL_CURSOR_ON_1 + "\n  Thinking… (12s · ↓ 1.1k tokens · esc to interrupt)"
    QUOTED = "\n".join([  # a reply that merely quotes the menu text — no live confirm footer highlight
        "● The menu said: 1. Stop and wait for limit to reset / 2. Upgrade your plan",
        "────────",
        "❯ ",
        "────────",
        "  ⏵⏵ bypass permissions on (shift+tab to cycle)",
    ])
    IDLE = "\n".join(["● done", "────────", "❯ ", "────────", "  ⏵⏵ bypass permissions on (shift+tab to cycle)"])
    PERM_EDIT = "\n".join([  # a Claude Code edit-approval gate
        "  Do you want to make this edit to config.ts?",
        "   ❯ 1. Yes",
        "     2. Yes, and don't ask again this session",
        "     3. No, and tell Claude what to do differently (esc)",
    ])
    PERM_BASH = "\n".join([  # a command-approval gate
        "  Do you want to run this command?",
        "     rm -rf /opt/claude/theoffice/tenant/store",
        "   ❯ 1. Yes",
        "     2. No, and tell Claude what to do differently (esc)",
    ])
    cases = [
        ("cursor on Stop-and-wait -> act", MODAL_CURSOR_ON_1, "act"),
        ("cursor on Upgrade -> REFUSE (never confirm Upgrade)", MODAL_CURSOR_ON_2, "refuse"),
        ("busy modal -> busy (never mid-turn)", MODAL_BUSY, "busy"),
        ("quoted menu in a reply -> none", QUOTED, "none"),
        ("idle pane -> none", IDLE, "none"),
        ("edit-permission prompt -> permission (alarm, never act)", PERM_EDIT, "permission"),
        ("bash-permission prompt -> permission (alarm, never act)", PERM_BASH, "permission"),
    ]
    bad = 0
    for name, pane, want in cases:
        got = classify(pane)
        ok = got == want
        print(f"[{'PASS' if ok else 'FAIL'}] {name}: got={got} want={want}")
        if not ok:
            bad += 1
    # Hard invariant: the ONLY verdict that ever sends a key is 'act'. Assert that NOTHING which would
    # spend money (Upgrade highlighted) or grant permission (any approval prompt) can be classified 'act'.
    for name, pane in [("cursor-on-Upgrade", MODAL_CURSOR_ON_2), ("edit-permission", PERM_EDIT),
                       ("bash-permission", PERM_BASH)]:
        if classify(pane) == "act":
            print(f"[FAIL] INVARIANT: {name} classified 'act' — could confirm a money/permission action")
            bad += 1
    print(f"\nself-test: {'ALL PASS' if bad == 0 else str(bad) + ' FAILED'}")
    return 1 if bad else 0


def main() -> int:
    if "--self-test" in sys.argv:
        return self_test()
    dry = "--dry-run" in sys.argv
    try:
        n = sweep(dry)
        if n:
            log(f"sweep complete: {n} modal(s) {'would be ' if dry else ''}dismissed")
    except Exception as ex:
        log(f"sweep ERROR: {ex!r}")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/session/modal-sweeper.ts
// Session-hygiene sweeper — the periodic backstop born from the 2026-08-04 delivery outage.
//
// Jobs the isReady-gated deliver path structurally cannot do, because a blocked pane makes the
// deliverer bail BEFORE it can act, and because some failures never touch the deliverer at all:
//
//   1. OWNER-DELIVERY WATCHDOG (the load-bearing one). An owner (source='channel') message must never go
//      unheard. A single STATE-OBSERVER alarms Szoszo when a channel row is EITHER (a) dropped
//      (status=failed) OR (b) still queued and undelivered past a threshold. (b) is the case the outage
//      actually hit — messages parked behind a modal, never attempted, so never 'failed' — which an
//      alarm keyed only on 'failed' would have missed entirely (Toby). It also backstops the modal
//      detectors failing open: if a future Claude Code changes the modal wording, nothing dismisses, the
//      pane reads generic 'unknown', delivery gates shut — and the age alarm still fires. The alarm is
//      PANE-INDEPENDENT (posted via the Slack bot to the owner, like auth-watchdog) because the failure
//      cause IS a broken pane; re-delivering into it is what already failed.
//
//   2. Dismiss the usage-limit modal — but ONLY when the highlight is provably on Stop-and-wait
//      (detectsUsageLimitModal). Any other arrangement is left untouched: the sweeper is structurally
//      incapable of selecting Upgrade (spending the owner's money).
//
//   3. Detect a permission/approval prompt and ALARM — NEVER dismiss it. Clearing one grants an
//      authorisation (delete/push/payment/deploy) blind. Freezing is the safe failure; a human decides.
//
//   4. Drop stale scheduler heartbeats so a moot briefing isn't delivered ahead of, or instead of, live
//      work. Owner and inter-agent messages are never age-dropped.

import type { EngineConfig } from "../types.js";
import { log } from "../logger.js";
import { listSessions, capturePane, sendKey } from "./tmux.js";
import {
  detectsUsageLimitModal,
  detectsUnsafeUsageLimitModal,
  detectsPermissionPrompt,
} from "./pane-state.js";
import {
  reapStaleScheduler,
  enqueueOutbound,
  listOwnerAlarmable,
  markOwnerAlarmed,
} from "../queue/index.js";
import { ownerDropAlarmText, ownerStallAlarmText, permissionPromptAlarmText } from "../queue/policy.js";

const logger = log("session-hygiene");

const TICK_MS = 20_000;
// Don't press Enter into the same session more than once per this window (guards a mid-render double-capture).
const DISMISS_COOLDOWN_MS = 15_000;
// Re-alarm about the same frozen permission prompt at most this often (a human is needed; don't page every tick).
const PERMISSION_REALARM_MS = 30 * 60 * 1000;

function envSec(name: string, fallback: number): number {
  const raw = process.env[name];
  if (raw == null || raw === "") return fallback;
  const n = Number(raw);
  return Number.isFinite(n) && n >= 0 ? n : fallback;
}

// A queued owner message not delivered within this many seconds is alarmed as "stuck". Well under the
// ~3.5h Szoszo waited on 2026-08-04. A legitimate long turn can trip it; the alarm is worded as
// "not yet delivered / may be stuck", which is honest either way.
const ownerStaleSec = () => envSec("OFFICE_OWNER_STALE_SEC", 15 * 60);
// Re-alarm a STILL-stuck queued owner message on this cadence (a dropped/failed one alarms once).
const ownerRealarmSec = () => envSec("OFFICE_OWNER_REALARM_SEC", 30 * 60);
// A queued scheduler heartbeat older than this is moot; drop it. Owner/bus/manual/system never dropped.
const staleSchedulerSec = () => envSec("OFFICE_STALE_SCHED_SEC", 2 * 60 * 60);

export function startSessionHygiene(cfg: EngineConfig): () => void {
  const socket = cfg.tmux.socket;
  let stopped = false;
  const lastDismissAt = new Map<string, number>();
  const lastPermAlarmAt = new Map<string, number>();

  // ---- 2 & 3: per-session modal handling ----
  const sweepModals = () => {
    for (const session of listSessions(socket)) {
      if (!session.startsWith("agent-")) continue; // only agent panes
      const pane = capturePane(socket, session);
      if (pane == null) continue;

      // 3. Permission/approval gate: DETECT + ALARM, never a keystroke. Clearing it would blind-approve.
      if (detectsPermissionPrompt(pane)) {
        const now = Date.now();
        if (now - (lastPermAlarmAt.get(session) ?? 0) >= PERMISSION_REALARM_MS) {
          lastPermAlarmAt.set(session, now);
          alarmOwner(cfg, permissionPromptAlarmText(session.replace(/^agent-/, ""), socket, session));
          logger.error({ session }, "permission/approval prompt detected — LEFT UNTOUCHED (a keystroke would blind-approve); alarmed owner");
        }
        continue;
      }

      // 2. Usage-limit modal present but NOT provably highlighted on Stop-and-wait: never press Enter
      // (could confirm Upgrade). Leave it and surface it.
      if (detectsUnsafeUsageLimitModal(pane)) {
        logger.error({ session }, "usage-limit modal present but highlight NOT on Stop-and-wait — left untouched (Enter could select Upgrade); needs a human");
        continue;
      }

      // 2. Usage-limit modal with the highlight provably on Stop-and-wait: safe to confirm option 1.
      if (!detectsUsageLimitModal(pane)) continue;
      const now = Date.now();
      if (now - (lastDismissAt.get(session) ?? 0) < DISMISS_COOLDOWN_MS) continue;
      lastDismissAt.set(session, now);
      const ok = sendKey(socket, session, "Enter");
      logger.warn({ session, ok }, "usage-limit modal (highlight on Stop-and-wait) — sent Enter to select it (auto-resumes at reset)");
    }
  };

  // ---- 1: owner-delivery watchdog ----
  const watchOwnerDelivery = () => {
    const rows = listOwnerAlarmable(ownerStaleSec(), ownerRealarmSec());
    for (const r of rows) {
      const channel = r.reply_channel ?? cfg.owner.slackUserId;
      if (!channel) {
        logger.error({ id: r.id, agent: r.agent_id }, "owner message needs escalation but NO reply channel / owner slack id — cannot alarm");
        continue;
      }
      const text = r.dropped
        ? ownerDropAlarmText(r.agent_id, `${r.status} after ${r.attempts} attempts`, r.prompt)
        : ownerStallAlarmText(r.agent_id, Math.round(r.age_sec / 60), r.prompt);
      try {
        enqueueOutbound(cfg.mainAgentId, channel, text);
        markOwnerAlarmed(r.id);
        logger.warn({ id: r.id, agent: r.agent_id, dropped: r.dropped, ageSec: r.age_sec }, "owner-delivery watchdog: alarmed owner (pane-independent)");
      } catch (err) {
        logger.error({ err, id: r.id }, "owner-delivery watchdog: failed to enqueue owner alarm (will retry next tick)");
      }
    }
  };

  const tick = () => {
    if (stopped) return;
    try { watchOwnerDelivery(); } catch (err) { logger.error({ err }, "owner-delivery watchdog tick failed"); }
    try { sweepModals(); } catch (err) { logger.error({ err }, "modal sweep failed"); }
    try {
      const dropped = reapStaleScheduler(staleSchedulerSec());
      if (dropped > 0) logger.warn({ dropped }, "dropped stale scheduler heartbeats (moot before delivery)");
    } catch (err) { logger.error({ err }, "stale-scheduler reap failed"); }
  };

  const handle = setInterval(tick, TICK_MS);
  logger.info({ tickMs: TICK_MS, ownerStaleSec: ownerStaleSec(), staleSchedulerSec: staleSchedulerSec() }, "session-hygiene sweeper started");
  return () => {
    stopped = true;
    clearInterval(handle);
  };
}

/** Pane-independent owner alarm: post to the owner via the Slack bot (bypasses the broken pane). */
function alarmOwner(cfg: EngineConfig, text: string): void {
  const channel = cfg.owner.slackUserId;
  if (!channel) {
    logger.error("no owner slack id configured — cannot raise a pane-independent alarm");
    return;
  }
  try {
    enqueueOutbound(cfg.mainAgentId, channel, text);
  } catch (err) {
    logger.error({ err }, "failed to enqueue owner alarm");
  }
}
